nanofi/examples/tail_file.c

```diff
@@ -0,0 +1,270 @@
+    if (!fp) return;
+    fseek(fp, curr_offset, SEEK_SET);
+    while (fgets(buff, max_bytes_read, fp) != NULL) {
+        set_offset(ftell(fp));
+        char * strings[max_bytes_read];
+        int num_strings = 0;
+        for (int i = 0; i < max_bytes_read; ++i) {
+            strings[i] = NULL;
+        }
+        split_string(buff, delim, strings, &num_strings);
+
+        flowfiles = (flow_file_records *)malloc(sizeof(flow_file_records));
+        flowfiles->records = malloc(sizeof(flow_file_record *) * num_strings);
+        flowfiles->len = num_strings;
+
+        for (int i = 0; i < num_strings; ++i) {
+            flowfiles->records[i] = NULL;
+        }
+
+        for (int i = 0; i < num_strings; ++i) {
+            flow_file_record * ffr = generate_flow_file(instance);
+            const char * flow_file = ffr->contentLocation;
+            FILE * ffp = fopen(flow_file, "wb");
+            if (!ffp) {
+                printf("Cannot open flow file at path %s to write content to.\n", flow_file);
+                fclose(fp);
+                free_all_strings(strings, num_strings);
+                return;
+            }
+            fwrite(strings[i], 1, strlen(strings[i]), ffp);
+            flowfiles->records[i] = ffr;
+            free(strings[i]);
+            fclose(ffp);
+        }
+    }
+    fclose(fp);
+}
+
+int main(int argc, char** argv) {
+
+    if (argc < 6) {
+        printf("Error: must run ./tail_file <file> <interval> <delimiter> <nifi instance url> <remote port>\n");
+        exit(1);
+    }
+
+    char * file = argv[1];
+    char * interval = argv[2];
+    char * delimiter = argv[3];
+    char * instance_str = argv[4];
+    char * port_str = argv[5];
+
+    if (access(file, F_OK) == -1) {
+        printf("Error: %s doesn't exist!\n", file);
+        exit(1);
+    }
+
+    struct stat stats;
+    stat(file, &stats);
+
+    // Check for file existence
+    if (S_ISDIR(stats.st_mode)){
+        printf("Error: %s is a directory!\n", file);
+        exit(1);
+    }
+
+    unsigned long intrvl = strtol(interval, NULL, 10);
+
+    if (errno == ERANGE || intrvl == LONG_MAX || intrvl == LONG_MIN) {
+        printf("Invalid interval value specified\n");
+        return 0;
+    }
+
+    nifi_port port;
+
+    port.port_id = port_str;
+
+    instance = create_instance(instance_str, &port);
+
+    const char * processor_name = "TailFile";
+
+    add_custom_processor(processor_name, on_trigger_callback);
+
+    standalone_processor *standalone_proc = create_processor(processor_name);
+
+    set_standalone_property(standalone_proc, "file_path", file);
+    set_standalone_property(standalone_proc, "delimiter", delimiter);
+
+    fileoffset = (file_offset *)malloc(sizeof(file_offset));
+    set_offset(0);
+    for (;;) {
+        flow_file_record * new_ff = invoke(standalone_proc);
+        transmit_flow_files(instance);
+        free_flowfile(new_ff);
+        sleep(intrvl);
+    }
+
+    free(fileoffset);
```

Comment:
I am not sure about how signal handling is implemented in nanofi (if at all), but I do not see how we could reach this code.

Reply:
I don't know about any handler for sigterm (which would apply here), so this is code is unreachable.